Jetpack Focus: show overlay on Jetpack feature card tap

zwarm opened the pull request:
Closes #17336

This PR hooks in the showing of the overlay view on tap of the dashboard card.
The logic is handled separately from the “screen_type” enum because this particular overlay did not fit that pattern. I am open to changing if the reviewer feels this would be a better implementation.


Light | Dark
-- | --
<img width="250" height="500" alt="Light" src="https://user-images.githubusercontent.com/506707/210891143-b2e80545-0b2d-4bbd-ba70-1958b061986c.png"> | <img width="250" height="500" alt="Dark" src="https://user-images.githubusercontent.com/506707/210891160-8e3585bf-18a5-4d42-b939-e9d06a63ff85.png">


The following existing events will be tracked with “source=card”.
remove_feature_overlay_displayed
remove_feature_overlay_dismissed
remove_feature_overlay_link_tapped
remove_feature_overlay_button_tapped

Notes:
I ran into an issue with the overlay not showing on subsequent taps of the card, this was due to `JetpackFeatureFullScreenOverlayFragment` using `ActivityViewModels`. I have switched this out to `viewModels`. It was definitely a head scratcher. I didn’t see any instances in which the viewModel was being shared, but please call me out if this is not the case.

**To test:**
- Install WP app
- Login
- Navigate to Me -> App Settings -> Privacy Settings and turn on Collection Information
- Navigate to -> Debug Settings -> and Enable jp_removal_three and restart the app
- Navigate to My Site
- ✅ Verify that the Jetpack Feature card is shown
- Tap on the card
- ✅ Verify the overlay is displayed
- ✅ Verify the log contains 🔵 Tracked: remove_feature_overlay_displayed, Properties: {"phase":"three","source":"card"}
- Tap on Continue with Jetpack
- ✅ Verify the overlay closes
- ✅ Verify the log contains 🔵 Tracked: remove_feature_overlay_dismissed, Properties: {"phase":"three","source":"card","dismissal_type":"continue"}
- Tap on the card
- Tap on Learn more at Jetpack.com link
- ✅ Verify the in-app browser launches
- Tap back to return to the app
- ✅ Verify the log contains 🔵 Tracked: remove_feature_overlay_link_tapped, Properties: {"phase":"three","source":"card"}
- ✅ Verify the overlay is not showing
- Tap on the card
- Tap on the "Switch to the Jetpack app" button
- ✅ Verify the log contains 🔵 Tracked: remove_feature_overlay_button_tapped, Properties: {"phase":"three","source":"card"}

## Regression Notes
1. Potential unintended areas of impact
The overlay is not shown on tap

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing

3. What automated tests I added (or what prevented me from doing so)
Update existing where needed

PR submission checklist:

- [X] I have completed the Regression Notes.
- [X] I have considered adding accessibility improvements for my changes.
- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

zwarm committed:
Add animations for all features overlay

zwarm committed:
Add strings for use on the all features overlay

zwarm committed:
Add source to OpenJetpackFeatureOverlay action

zwarm committed:
Update newInstance to include flags for the all features overlay

zwarm committed:
Address detekt issues

zwarm committed:
Add methods for tracking all feature overlay along with a new enum to capture the source.

zwarm committed:
Add logic to launch the overlay

zwarm committed:
Add AllFeaturesPhase sealed classes for assisting with showing the all feature overlay

zwarm committed:
Refactor: Add support for the all feature overlay

zwarm committed:
Refactor: address detekt warnings

zwarm committed:
Refactor: use viewModels instead of activityViewModels

zwarm committed:
Refactor: remove uneeded import statement

zwarm committed:
Update AllFeaturesPhase.PhaseThree to include migrationText and InfoText

zwarm committed:
Refactor: Set values for migrationText and InfoText

zwarm committed:
Implement event tracking for all features overlay

zwarm committed:
Fix detekt line too long

zwarm committed:
Refactor: change label from feature_card to card

zwarm committed:
Remove test

AjeshRPai committed:
Updates: the Overlay viewmodel and fragment

Updates: the viewmodel fetching logic and the event type of the overlay
When the viewmodel is fetched using viewmodels instead of the activity
viewmodels, the site creation logic will fail. Making the events single
 live event would ensure that these events are consumed only once

AjeshRPai reviewed (approved):
Hey @zwarm 

I went through the PR and tested according to instructions, and Everything works as expected. But the change to `JetpackFeatureFullScreenOverlayFragment` would affect the site creation overlay logic in Site creation Phase 2. In site creation phase 2, the site creation is disabled. 

Changing viewModel fetching logic from 
 `private val viewModel: JetpackFeatureFullScreenOverlayViewModel by activityViewModels()
`
to 

` private val viewModel: JetpackFeatureFullScreenOverlayViewModel by viewModels()`

Would fail the following scenario. 

- Go to app and login 
- Make sure the app is in site creation phase 2 (`jp_removal_four` is enabled)
- Try to create a site 
- Verify that the Site creation overlay is shown 
- Dismissing the overlay should take you back to site switcher page 

In this case, the site creation activity is dismissed by observing the dismiss action of the overlay in ViewModel. When the ViewModel is not created with the actvitiyViewmodels() the ViewModel instance between the SiteCreationActivity and the `JetpackFeatureFullScreenOverlayFragment` is not the same. The Site creation activity will be seen even if the Overlay fragment is dismissed. 

While testing I found that the card tapping doesn't show the overlay when the ViewModel is fetched using `activityViewmodels()` in `JetpackFeatureFullScreenOverlayFragment`. On further testing, I found that this is not due to how ViewModel is created but because the overlay events are `LiveData` and not `SingleLiveEvent`. When I changed the events in overlay fragments from `LiveData` to `SingleLiveEvent`, also changed the ViewModel creation logic back to `private val viewModel: JetpackFeatureFullScreenOverlayViewModel by activityViewModels()`, i found that its working as expected. 

Video capture before converting the Overlay events to `SingleLiveEvent` but the ViewModel logic is  `private val viewModel: JetpackFeatureFullScreenOverlayViewModel by activityViewModels()`. Here you can see that repeatedly tapping the card will show the overlay but once we dismiss the overlay using any events other than the back navigation, the overlay won't be shown again. 

https://user-images.githubusercontent.com/17463767/211010166-c9d2fc4a-951a-4bf6-baa8-ed6fadfe7cde.mp4

After changing the event type of Overlay from `LiveData` to `SingleLiveEvent` with commit - 93fbd36d0d34b3998f86d421f708869a0ef955c8. Video capture 👇🏼 



https://user-images.githubusercontent.com/17463767/211010391-b03d9f58-fc0b-41cb-9281-2845ee28f088.mp4

Am approving but not merging the PR so that you can have a look and confirm the changes I have done makes sense. Other than this change, I found everything is working as expected 👍🏼

AjeshRPai committed:
Removes: redundant imports

zwarm commented:
@AjeshRPai - Thanks!

> I went through the PR and tested according to instructions, and Everything works as expected. But the change to JetpackFeatureFullScreenOverlayFragment would affect the site creation overlay logic in Site creation Phase 2.

That's what I was afraid of ... there was something I wasn't seeing. Thanks for remembering!

> On further testing, I found that this is not due to how ViewModel is created but because the overlay events are LiveData and not SingleLiveEvent. When I changed the events in overlay fragments from LiveData to SingleLiveEvent, also changed the ViewModel creation logic back to private val viewModel: JetpackFeatureFullScreenOverlayViewModel by activityViewModels(), i found that its working as expected.

I see, I see. Using `SingleLiveEvent` makes sense since the updated value is sent only once. This fits our scenario. 🙇 Thank you!!!
